### PR TITLE
Issue 1324 netcdf parallel

### PR DESCRIPTION
**Pull request by doutriaux1**

### CODE REVIEW

**Comment by doutriaux1** [comment]
@dlonie @aashish24 this PR has evolved into something much diff than its original name.

It does turn on PARALLEL capabilities for both hdf5/netcdf4 and also cdms2.

BUT cdms2 // writing is still very much in development, // writing must be done "the right" so while the features are here for experts, the newbie will likely break its code. I will keep exploring the // aspect and improve both cdms2 and documentation on how to use it in this context.

What this PR is really about is
1- Many build fixes on various Lniux, most difficult one being rhea
2- newer NetCDF that gave us 2 order of magnitude speed gain on rhea (along with some configuration  tweaks withini rhea)
3: 0d write bug fix in cdms2, string array bug fixes.

all of this is VERY important and high priority for ACME, alsofixes master build on linux rh and ubunutu for me.

possible side effect: loss of dap client (bug in the newer netcdf) I'm investigating this

**Comment by aashish24** [comment]
@doutriaux1 one minor request. Can you squash some of the commits such as "getting there" so that history is bit cleaner and its easy to understand your changes. Also, we may need to get rid of building ffi (if not in this branch then in another PR when we update the add_cdat_package and deps cmake macros).

**Comment by doutriaux1** [comment]
@aashish24 found a bug (ctest thank you) so put it in now. Let's keep ffi in, when we you use_system_packages it will just skip it, otherwise people like susannah with no admin won't be able to build it.

**Comment by doutriaux1** [comment]
@aashish24 there's only one "getting there" and it's pretty descriptive there. I promise to work even harder on better commit messages in the future. Feel free to squash commits to your liking.

**Comment by doutriaux1** [comment]
@aashish24 do you know of a quick tool to squash commits?

**Comment by doutriaux1** [comment]
looking into test1 failure

**Comment by doutriaux1** [comment]
ok last not dap related test passes see mac results at:
https://open.cdash.org/buildSummary.php?buildid=3859039

**Comment by dlonie** [comment]
Squashing commits: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

`git rebase -i [commitish]` is the easiest way. Takes less than a minute for a simple squash.

**Comment by aashish24** [comment]
thanks @doutriaux1 , you can use command referred by @dlonie. You can pick a commit that you want to use as a root commit (from that point you want to change history) and then it will show a list of commits  something like

pick ff225e6 Fix #1029 - Update ffmpeg from 0.11.1 to 2.7
pick 1e21fec Add whitespace to trigger rebuild

You can change pick to s for squash. You cannot squash top commit in that list if that makes sense.

**Comment by doutriaux1** [comment]
@dlonie yep I know that but it requires some thinking :stuck_out_tongue_closed_eyes: I was hoping for a gui-based once (I know I must be getting old when I ask for a GUI tool...) for less brain power drain.

**Comment by dlonie** [comment]
It's really quite simple -- run the command, change `pick` to `f`(ixup) (or `s`(quash) depending on whether you want to keep the second commit message), and it's good to go. Won't even have to worry about conflicts unless you rearrange the commits.

**Comment by doutriaux1** [comment]
@dlonie it's no problem. Plus it's easier now that I'm at work, got the bigger screens can see multiple windows at once etc... Will squash in a minute.

**Comment by doutriaux1** [comment]
@aashish24 is it better now?

**Comment by aashish24** [comment]
Its looking much better I think. It could be improved bit more but for now it will work.

File: CMake/cdat_modules/netcdf_external.cmake [hunk 1]
```diff
@@ -8,16 +9,23 @@ configure_file(${cdat_CMAKE_SOURCE_DIR}/cdat_modules_extra/netcdf_patch_step.cma
...
   URL_MD5 ${NC4_MD5}
-  BUILD_IN_SOURCE 1
+  #${NC4_GIT_CMD}
+  #${NC4_GIT_TAG}
+  CMAKE_CACHE_ARGS
+       ${netcdf_build_args}
+  CMAKE_ARGS
+    -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
   PATCH_COMMAND ${netcdf_PATCH_COMMAND}
-  CONFIGURE_COMMAND ${CMAKE_COMMAND} -DINSTALL_DIR=<INSTALL_DIR> -DWORKING_DIR=<SOURCE_DIR> -D CONFIGURE_ARGS=${netcdf_configure_args} -P ${cdat_CMAKE_BINARY_DIR}/cdat_configure_step.cmake
```
Comment (aashish24) [inline CMake/cdat_modules/netcdf_external.cmake]: Why you took out this?

File: CMake/cdat_modules_extra/cdatmpi_configure_step.cmake.in [hunk 2]
```diff
@@ -0,0 +1,21 @@
+set(ENV{PKG_CONFIG_PATH} "@cdat_EXTERNALS@/lib/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig:/usr/lib64/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:$ENV{PKG_CONFIG_PATH}")
+set(ENV{PKG_CONFIG} "@cdat_PKG_CONFIG_EXECUTABLE@")
+
+include(@cdat_CMAKE_BINARY_DIR@/cdat_common_environment.cmake)
+
+set(ENV{CC} mpicc)
+
+message("CONFIGURE_ARGS IS ${CONFIGURE_ARGS}")
```
Comment (aashish24) [inline CMake/cdat_modules_extra/cdatmpi_configure_step.cmake.in]: Please take out this debug messages

**Comment by aashish24** [comment]
@doutriaux1 looks like things are failing in this branch. Do you know why?

**Comment by aashish24** [comment]
Successfully updated your environment to use UVCDAT
(changes are valid for this session/terminal only)
Version: 2.2.0-99-gb7f1a56
Location: /home/kitware/buildbot/buildbot-slave/uvcdat-test-laptop-linux-release/build/install
Traceback (most recent call last):
  File "/home/kitware/buildbot/buildbot-slave/uvcdat-test-laptop-linux-release/source/testing/vcs/test_vcs_basic_vectors.py", line 25, in <module>
    import vcs
  File "/home/kitware/buildbot/buildbot-slave/uvcdat-test-laptop-linux-release/build/install/lib/python2.7/site-packages/vcs/__init__.py", line 32, in <module>
    from utils import *
  File "/home/kitware/buildbot/buildbot-slave/uvcdat-test-laptop-linux-release/build/install/lib/python2.7/site-packages/vcs/utils.py", line 3, in <module>
    import cdtime
ImportError: No module named cdtime

File: CMake/cdat_modules_extra/cdatmpi_configure_step.cmake.in (hunk 2 above)
Comment (doutriaux1) [inline CMake/cdat_modules_extra/cdatmpi_configure_step.cmake.in]: :+1:  thanks for spotting and reviewing!

**Comment by doutriaux1** [comment]
not sure works for me on mac(s), rhea, ubnutu.

**Comment by doutriaux1** [comment]
@aashish24 I will do a fresh build just to be sure.

**Comment by doutriaux1** [comment]
@aashish24 I got it the switch to CMake build seems to have changed where libnetcdf goes... it's now (on my ubuntu at least) under: 
```
./Externals/lib/x86_64-linux-gnu/libnetcdf.so.7
./Externals/lib/x86_64-linux-gnu/libnetcdf.settings
./Externals/lib/x86_64-linux-gnu/libnetcdf.so.7.2.1
./Externals/lib/x86_64-linux-gnu/libnetcdf.so
```
Will need to go back to mac to double check what happen there

**Comment by doutriaux1** [comment]
ok on my mac it's at the "right" place Externals/lib... Will revert the change for CMake build

**Comment by aashish24** [comment]
@doutriaux1 bunch of tests are still waiting. Mind looking into it?

**Comment by doutriaux1** [comment]
@aashish24 I know I fixed this somewhere... Probably didn't push the branch, will push soon.

**Comment by doutriaux1** [comment]
@aashish24 seems to pass on garant, no idea what the test failure is about but I don't think it has anything to do with netcdf :wink:

**Comment by doutriaux1** [comment]
@dlonie I also found another bunch of object that are preciously kept by vcs forever in #1424 fixing this might help you too.

**Comment by durack1** [comment]
@doutriaux1 the [issue with garant](https://open.cdash.org/testDetails.php?test=347749390&build=3873138) can be fixed with the `sleep(1)` added in #1393 - a rebase should solve that..

**Comment by aashish24** [comment]
thanks @doutriaux1 dashboards are looking pretty good now.

**Comment by aashish24** [comment]
:+1: I didn't verify the parallel code but in general it looks good to me.

**Pull request merged**